hack/setup.sh

```diff
@@ -0,0 +1,129 @@
+#!/bin/bash
+
+# Exit immediately if a command exits with a non-zero status.
+set -e
+
+echo "Installing Knative Serving Custom Resources..."
+kubectl apply -f https://github.com/knative/serving/releases/download/knative-v1.12.3/serving-crds.yaml
+
+echo "Installing Knative Serving Core Components..."
+kubectl apply -f https://github.com/knative/serving/releases/download/knative-v1.12.3/serving-core.yaml
+
+echo "Installing Knative Kourier Network Layer..."
+kubectl apply -f https://github.com/knative/net-kourier/releases/download/knative-v1.12.3/kourier.yaml
+
+echo "Configuring Knative Serving to use Kourier..."
+kubectl patch configmap/config-network \
+  --namespace knative-serving \
+  --type merge \
+  --patch '{"data":{"ingress-class":"kourier.ingress.networking.knative.dev"}}'
+
+echo "Fetching External IP or CNAME..."
+kubectl --namespace kourier-system get service kourier
+
+echo "Waiting for all Knative Serving pods to be in the Running state..."
+
+while IFS= read -r line; do
+    POD_NAME=$(echo $line | awk '{print $1}')
+    POD_STATUS=$(echo $line | awk '{print $3}')
+    if [ "$POD_STATUS" != "Running" ]; then
+        echo "Waiting for pod $POD_NAME to be in Running state..."
+        sleep 5
+        kubectl get pods --namespace knative-serving | grep -v NAME | while IFS= read -r line; do continue; done
+    fi
+done < <(kubectl get pods --namespace knative-serving | grep -v NAME)
+
+echo "All pods are running."
+
+echo "Verifying Knative Serving Installation..."
+kubectl get pods --namespace knative-serving
+
```

Comment:
What's the point of this if we had the previous loop check that all pods are running?